File: matlab/testLin2.m
% LIN2 Problem Test

clear; clc;

% Define objective function
lin2Objective = @(x) -x(1) - x(2);

% Define constraints (inverted from >= 0 to <= 0)
g1 = @(x) -(2*x(1)^4 - 8*x(1)^3 + 8*x(1)^2 - x(2) + 2);
g2 = @(x) -(4*x(1)^4 - 32*x(1)^3 + 88*x(1)^2 - 96*x(1) - x(2) + 36);

% Variable bounds
lin2Bounds = [0.0 3.0; 0.0 4.0];

% Number of runs
runsCount = 100;
executionValues = zeros(runsCount, 1);
executionFevals = zeros(runsCount, 1);

fprintf('Starting validation: Running LIN2 optimization %d times...\n', runsCount);

for run = 1:runsCount
    solver = ConstrainedGA('ObjectiveFunction', lin2Objective, ...
                           'InequalityConstraints', {g1, g2}, ...
                           'Bounds', lin2Bounds, ...
                           'PopulationSize', 200, ...
                           'MaxGenerations', 200, ...
                           'PenaltyMultiplier', 1000.0);
    
    runResult = solver.solve();
    executionValues(run) = runResult.value;
    executionFevals(run) = runResult.fevalsSpent;
    
    if mod(run, 20) == 0
        fprintf('  > Completed %d/%d runs...\n', run, runsCount);
    end
end

% Compute statistics
avgFevals = mean(executionFevals);
bestValue = min(executionValues);
meanValue = mean(executionValues);
stdDeviation = std(executionValues);

% Display results
fprintf('\n=======================================================\n');
fprintf('           LIN2 PROBLEM BENCHMARK RESULTS              \n');
fprintf('=======================================================\n');
fprintf('Metric       |  Paper Table 2 Data  | MATLAB Run\n');
fprintf('-------------------------------------------------------\n');
fprintf('FEVALS       |  2431                | %d\n', round(avgFevals));
fprintf('BEST         |  -5.5080             | %.4f\n', bestValue);
fprintf('MEAN         |  -5.5080             | %.4f\n', meanValue);
fprintf('STD          |  2.17e-1             | %.2e\n', stdDeviation);
fprintf('=======================================================\n');

File: javascript/testLin2.js
const ConstrainedGA = require('./constrainedGA');

// Section 3.1: LIN2 Problem Setup
// min f(x) = -x1 - x2
const lin2Objective = (x) => -x[0] - x[1];

// Inverting paper's >= 0 constraints to meet our engine's <= 0 input format
const g1 = (x) => -(2 * Math.pow(x[0], 4) - 8 * Math.pow(x[0], 3) + 8 * Math.pow(x[0], 2) - x[1] + 2);
const g2 = (x) => -(4 * Math.pow(x[0], 4) - 32 * Math.pow(x[0], 3) + 88 * Math.pow(x[0], 2) - 96 * x[0] - x[1] + 36);

const lin2Bounds = [
    [0.0, 3.0], // x1 bounds
    [0.0, 4.0]  // x2 bounds
];

const runsCount = 100;
const executionValues = [];
const executionFevals = [];

console.log(`Starting validation: Running LIN2 optimization ${runsCount} times...`);

for (let run = 1; run <= runsCount; run++) {
    const solver = new ConstrainedGA({
        objFunc: lin2Objective,
        inequalityConstraints: [g1, g2],
        bounds: lin2Bounds,
        K: 200,
        iterMax: 200,
        lam: 1000.0
    });

    const runResult = solver.solve();
    executionValues.push(runResult.value);
    executionFevals.push(runResult.fevalsSpent);
    
    if (run % 20 === 0) {
        console.log(`  > Completed ${run}/${runsCount} runs...`);
    }
}

// Compute statistics
const avgFevals = executionFevals.reduce((a, b) => a + b, 0) / runsCount;
const bestValue = Math.min(...executionValues);
const meanValue = executionValues.reduce((a, b) => a + b, 0) / runsCount;
const variance = executionValues.reduce((sum, val) => sum + Math.pow(val - meanValue, 2), 0) / runsCount;
const stdDeviation = Math.sqrt(variance);

// Benchmark Table Output
console.log("\n=======================================================");
console.log("           LIN2 PROBLEM BENCHMARK RESULTS              ");
console.log("=======================================================");
console.log(`Metric       |  Paper Table 2 Data  | Your Local Run JS`);
console.log("-------------------------------------------------------");
console.log(`FEVALS       |  2431               | ${Math.round(avgFevals)}`);
console.log(`BEST         |  -5.5080            | ${bestValue.toFixed(4)}`);
console.log(`MEAN         |  -5.5080            | ${meanValue.toFixed(4)}`);
console.log(`STD          |  2.17e-1            | ${stdDeviation.toExponential(2)}`);
console.log("=======================================================");